Resolve: VectorSearch enabled SQLChain?

mpskex opened the pull request:
<!-- Thank you for contributing to LangChain!

Replace this comment with:
  - Description: a description of the change, 
  - Issue: the issue # it fixes (if applicable),
  - Dependencies: any dependencies required for this change,
  - Tag maintainer: for a quicker response, tag the relevant maintainer (see below),
  - Twitter handle: we announce bigger features on Twitter. If your PR gets announced and you'd like a mention, we'll gladly shout you out!

If you're adding a new integration, please include:
  1. a test for the integration, preferably unit tests that do not rely on network access,
  2. an example notebook showing its use.

Maintainer responsibilities:
  - General / Misc / if you don't know who to tag: @baskaryan
  - DataLoaders / VectorStores / Retrievers: @rlancemartin, @eyurtsev
  - Models / Prompts: @hwchase17, @baskaryan
  - Memory: @hwchase17
  - Agents / Tools / Toolkits: @hinthornw
  - Tracing / Callbacks: @agola11
  - Async: @agola11

If no one reviews your PR within a few days, feel free to @-mention the same people again.

See contribution guidelines for more information on how to write/run tests, lint, etc: https://github.com/hwchase17/langchain/blob/master/.github/CONTRIBUTING.md
 -->

Hello from [MyScale](https://myscale.com/) AI team! 😊👋

We have been working on features to fill up the gap among SQL, vector search and LLM applications. Some inspiring works like self-query retrievers for VectorStores (for example [Weaviate](https://python.langchain.com/en/latest/modules/indexes/retrievers/examples/weaviate_self_query.html) and [others](https://python.langchain.com/en/latest/modules/indexes/retrievers/examples/self_query.html)) really turn those vector search databases into a powerful knowledge base! 🚀🚀

We are thinking if we can merge all in one, like SQL and vector search and LLMChains, making this SQL vector database memory as the only source of your data. Here are some benefits we can think of for now, maybe you have more 👀:

With ALL data you have: since you store all your pasta in the database, you don't need to worry about the foreign keys or links between names from other data source.
Flexible data structure: Even if you have changed your schema, for example added a table, the LLM will know how to JOIN those tables and use those as filters.
SQL compatibility: We found that vector databases that supports SQL in the marketplace have similar interfaces, which means you can change your backend with no pain, just change the name of the distance function in your DB solution and you are ready to go!

### Issue resolved:
- [Feature Proposal: VectorSearch enabled SQLChain?](https://github.com/hwchase17/langchain/issues/5122)

### Change made in this PR:
- An improved schema handling that ignore `types.NullType` columns 
- A SQL output Parser interface in `SQLDatabaseChain` to enable Vector SQL capability and further more
- A Retriever based on `SQLDatabaseChain` to retrieve data from the database for RetrievalQAChains and many others
- Allow `SQLDatabaseChain` to retrieve data in python native format
- Includes PR #6737 
- Vector SQL Output Parser for `SQLDatabaseChain` and `SQLDatabaseChainRetriever`
- Prompts that can implement text to VectorSQL
- Corresponding unit-tests and notebook

### Twitter handle: 
- @MyScaleDB

### Tag Maintainer:
Prompts / General: @hwchase17, @baskaryan
DataLoaders / VectorStores / Retrievers: @rlancemartin, @eyurtsev

### Dependencies:
No dependency added

Commit:
constructor /  /  implementation

mpskex committed:
Merge branch 'hwchase17:master' into feature/vectorstore_myscale

Commit:
update

Commit:
update impl

mpskex committed:
Merge branch 'hwchase17:master' into feature/vectorstore_myscale

Commit:
change default

mpskex committed:
Merge branch 'hwchase17:master' into feature/vectorstore_myscale

Commit:
add unit tests

add unit tests

revised

mpskex committed:
Merge branch 'master' into feature/vectorstore_myscale

Commit:
unittest passed

Commit:
add escape string

Commit:
format / lint

mpskex committed:
update poetry.lock

Commit:
revised `__repr__` docstrings

mpskex committed:
Merge branch 'master' into feature/vectorstore_myscale

Commit:
Merge pull request #1 from myscale/feature/vectorstore_myscale

Feature/vectorstore myscale

mpskex committed:
change `tqdm` into an optional package

Commit:
fixed typo in myscale docs

mpskex committed:
Merge branch 'hwchase17:master' into master

Commit:
revised prompted operator names

Commit:
improve sqlalchemy support

mpskex committed:
Merge branch 'hwchase17:master' into myscale/sql_self_query

Commit:
add sql_cmd_parser and examples

Commit:
extensive for other VectorSQL

Commit:
update interface to output parser

Commit:
Merge branch 'myscale/improve_string_pattern_match' into preview

mpskex committed:
Merge branch 'hwchase17:master' into preview

mpskex committed:
Merge branch 'hwchase17:master' into myscale/sql_self_query

Commit:
add sql database retriever

Commit:
revised sql database chain retriever

Commit:
Merge branch 'preview' into myscale/sql_self_query

Commit:
lint and format and tests

Commit:
format and fix BaseRow import

Commit:
add aget method

mpskex committed:
Merge branch 'hwchase17:master' into myscale/sql_self_query

Commit:
revised docs

Commit:
revised sql self query notebook

mpskex committed:
Merge branch 'hwchase17:master' into myscale/sql_self_query

Commit:
format and lint

baskaryan commented on langchain/chains/sql_database/base.py:
```diff
@@ -50,6 +50,10 @@ class SQLDatabaseChain(Chain):
     to fix the initial SQL from the LLM."""
     query_checker_prompt: Optional[BasePromptTemplate] = None
     """The prompt template that should be used by the query checker"""
+    sql_cmd_parser: Optional[BaseOutputParser] = None
```
hm we can already add an output parser to the llm chain, would that work instead?

mpskex replied:
Extracting embedding for words takes too much vectors and query checker will not be able to check that long sql. Putting the parser here will make the chain have direct control of the parser.

Also, for better readability we think dumping direct result from LLM is better than the parsed one.

For example 
```sql
SELECT * FROM table ORDER BY DISTANCE(vector, NeuralArray(langchain)) LIMIT 5
```
Or you have to dealing with
```sql
SELECT * FROM table ORDER BY distance(vector, [0.2314, 0.132415, ..., 0.13941] LIMIT 5
```
which does not make any sense to human.

mpskex replied:
There is another implementation can inject parsers into the llm_chain:
1. define a `SQLCommand` object to store both the parsed and unparsed llm output
2. force the llm_chains to always produce a SQLCommand object for further use of the raw output from LLM

This is a probable approach as far as I can think of. Which one do you prefer?

mpskex replied:
Revised with a dictionary that preserved output from LLM.

baskaryan commented on langchain/chains/query_constructor/ir.py:
```diff
@@ -75,8 +75,8 @@ class Comparator(str, Enum):
-    CONTAIN = "contain"
-    LIKE = "like"
+    CONTAIN = "list_contain"
+    LIKE = "string_pattern_like"
```
slight concern that this will change default behavior for existing query constructor chains. wonder if we should leave enum as is but make it easier to modify the name that's actually passed into the prompt when the query constructor chain is initialized?

mpskex replied:
Yes this is an issue we found when playing with self-query. We have opened a PR #6737 to fix this.

The major issue is that the CONTAIN comparator sometimes takes place on non-array columns and LIKE comparator sometimes appears on non-string columns. I do agree that this may change the behavior. but since these two comparators are new and added in #6143, I do think this fix may improve the prompt stability and all new features built on this may be enhanced.

Or we can leave this unchanged and open another PR for dynamic prompt for comparator?

mpskex replied:
Reverted back and wait for future dynamic prompt PR.

baskaryan commented on langchain/retrievers/sql_database.py:
```diff
@@ -0,0 +1,42 @@
+        run_manager: CallbackManagerForRetrieverRun,
+        **kwargs: Any,
+    ) -> List[Document]:
+        if not self.sql_db_chain.native_format:
```
probably something that could be checked at instantiation

mpskex replied:
Revised with validator to check at instantiation.

baskaryan reviewed (commented):
will admit, pr is a bit hard to review just because of how much is being added 😅

Commit:
revised according to comments in #7454

mpskex committed:
Merge branch 'hwchase17:master' into myscale/sql_self_query

Commit:
lint and fix type

Commit:
revert back comparator prompt

mpskex committed:
Merge branch 'hwchase17:master' into myscale/sql_self_query

mpskex committed:
Merge branch 'master' into myscale/sql_self_query

mpskex committed:
Merge branch 'master' into myscale/sql_self_query

mpskex committed:
revert back & relint & reformat

revert

revised to experimentals

try to resolve conflict

reverse change

try to resolve conflict

revised unittest

reformat & relint

revert

update poetry

mpskex commented:
Got several cells conflicting in `langchain/docs/extras/modules/data_connection/retrievers/how_to/self_query/myscale_self_query.ipynb` 

## SQL Self-Query Retriever with MyScale

```python3
!pip3 install clickhouse-sqlalchemy InstructorEmbedding sentence_transformers openai langchain-experimental
```

```python3
from os import environ
import getpass
from typing import Dict, Any
from langchain import OpenAI, SQLDatabase, SQLDatabaseChain, LLMChain
from sqlalchemy import create_engine, Column, MetaData
from langchain import PromptTemplate


from sqlalchemy import create_engine

MYSCALE_HOST = "msc-1decbcc9.us-east-1.aws.staging.myscale.cloud"
MYSCALE_PORT = 443
MYSCALE_USER = "chatdata"
MYSCALE_PASSWORD = "myscale_rocks"
OPENAI_API_KEY = getpass.getpass("OpenAI API Key:")

engine = create_engine(
    f"clickhouse://{MYSCALE_USER}:{MYSCALE_PASSWORD}@{MYSCALE_HOST}:{MYSCALE_PORT}/default?protocol=https"
)
environ["OPENAI_API_KEY"] = OPENAI_API_KEY
```

```python3
from langchain.embeddings import HuggingFaceInstructEmbeddings
from langchain_experimental.sql.parser import VectorSQLOutputParser

output_parser = VectorSQLOutputParser.from_embeddings(
    model=HuggingFaceInstructEmbeddings(
        model_name="hkunlp/instructor-base", model_kwargs={"device": "cpu"}
    )
)
```

```python3
from langchain.callbacks import StdOutCallbackHandler

metadata = MetaData(bind=engine)

from langchain.llms import OpenAI
from langchain.chat_models import ChatOpenAI

from langchain_experimental.utilities.sql_database import SQLDatabase
from langchain_experimental.sql.prompt import MYSCALE_PROMPT
from langchain_experimental.sql.base import SQLDatabaseChain

chain = SQLDatabaseChain(
    llm_chain=LLMChain(
        llm=OpenAI(openai_api_key=OPENAI_API_KEY, temperature=0),
        prompt=MYSCALE_PROMPT,
        output_parser=output_parser,
    ),
    top_k=10,
    return_direct=True,
    database=SQLDatabase(engine, None, metadata),
    native_format=True,
)

import pandas as pd

pd.DataFrame(
    chain.run(
        "Please give me 10 papers to ask what is PageRank?",
        callbacks=[StdOutCallbackHandler()],
    )
)
```

## SQL Database as Retriever

```python3
from langchain.chains.qa_with_sources.retrieval import RetrievalQAWithSourcesChain

from langchain_experimental.retrievers.sql_database import SQLDatabaseChainRetriever
from langchain_experimental.sql.parser import VectorSQLRetrieveAllOutputParser

output_parser_retrieve_all = VectorSQLRetrieveAllOutputParser.from_embeddings(
    output_parser.model
)

chain = SQLDatabaseChain.from_llm(
    llm=OpenAI(openai_api_key=OPENAI_API_KEY, temperature=0),
    prompt=MYSCALE_PROMPT,
    top_k=10,
    return_direct=True,
    db=SQLDatabase(engine, None, metadata),
    sql_cmd_parser=output_parser_retrieve_all,
    native_format=True,
)

# You need all those keys to get docs
retriever = SQLDatabaseChainRetriever(sql_db_chain=chain, page_content_key="abstract")

document_with_metadata_prompt = PromptTemplate(
    input_variables=["page_content", "id", "title", "authors", "pubdate", "categories"],
    template="Content:\n\tTitle: {title}\n\tAbstract: {page_content}\n\tAuthors: {authors}\n\tDate of Publication: {pubdate}\n\tCategories: {categories}\nSOURCE: {id}",
)

chain = RetrievalQAWithSourcesChain.from_chain_type(
    ChatOpenAI(
        model_name="gpt-3.5-turbo-16k", openai_api_key=OPENAI_API_KEY, temperature=0.6
    ),
    retriever=retriever,
    chain_type="stuff",
    chain_type_kwargs={
        "document_prompt": document_with_metadata_prompt,
    },
    return_source_documents=True,
)
ans = chain("Please give me 10 papers to ask what is PageRank?")
print(ans["answer"])
```

Commit:
add dependency

Commit:
Merge branch 'master' into myscale/sql_self_query

mpskex commented:
I think we are pretty close now. I moved everything to experimental. Could you please review this? @hwchase17

mpskex committed:
Merge branch 'langchain-ai:master' into myscale/sql_self_query

chtlp commented:
Some of our customers are requiring this feature, @hwchase17 please let us know how to make this feature get through.

mpskex committed:
Merge branch 'langchain-ai:master' into myscale/sql_self_query

hwchase17 reviewed (commented):
sorry for the delayed review

since this touches the sql/base file its going to be a bit harder to review than is probably worth it

the easiest way to move forward is probably to make this a completely standalone module in experimental. eg no changes to the existing files. i dont entirely understand the changes, but either way they should probably be done in a separate pr

an end-to-end example notebook would also be helpful here

mpskex commented:
Thanks for the comment! I have already migrated this into experimental. And I should have added a notebook to demonstrate what it is doing. I will be working on it now. Thanks for this again ❤️

mpskex committed:
Merge branch 'master' into myscale/sql_self_query

Commit:
lint & add notebook

mpskex commented:
> sorry for the delayed review
> 
> since this touches the sql/base file its going to be a bit harder to review than is probably worth it
> 
> the easiest way to move forward is probably to make this a completely standalone module in experimental. eg no changes to the existing files. i dont entirely understand the changes, but either way they should probably be done in a separate pr
> 
> an end-to-end example notebook would also be helpful here

I just added a notebook here: `docs/extras/modules/data_connection/retrievers/sql_database/myscale_vector_sql.ipynb`

Commit:
fixed lint

mpskex committed:
Merge branch 'master' into myscale/sql_self_query

mpskex committed:
update poetry.lock

mpskex committed:
move it out of sqldatabasechain

mpskex commented:
Close because we squashed commits into new -> PR #10177
